### PR TITLE
feat: 652 python template pubsublite to bigtable

### PR DESCRIPTION
**Pull request by tanyarw**

### CODE REVIEW

**Comment by vanshaj-bhatia** [comment]
@tanyarw please add integration test to `python/.ci/Jenkinsfile`

**Review by NiloFreitas — CHANGES_REQUESTED** [review]
Hi Tanya! Great work!! I left some comments with some refactor suggestions.
Cheers

File: python/dataproc_templates/pubsublite/pubsublite_to_bigtable.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            .load()
+
+        input_data = input_data.withColumn("data", input_data.data.cast(StringType()))
+
+        # Write
+        options = {}
+        if checkpoint_location:
+            options = {constants.PUBSUBLITE_CHECKPOINT_LOCATION: checkpoint_location}
+
+        def write_to_bigtable(batch_df: DataFrame, batch_id: int):
```
[the remark this answers is not in the document]
Comment (tanyarw) [reply python/dataproc_templates/pubsublite/pubsublite_to_bigtable.py:213]: writeStream.foreachBatch has this definition
`def foreachBatch(func: (DataFrame, int) -> None) -> DataStreamWriter`
I'm not sure how to exclude batch_id yet

File: python/.ci/Jenkinsfile [hunk 2]
```diff
@@ -430,6 +430,57 @@ pipeline {
...
+                                        cd python
+
+                                        ./bin/start.sh \
+                                        -- --template=PUBSUBLITETOBIGTABLE \
+                                        --pubsublite.bigtable.subscription.path=projects/$GCP_PROJECT/locations/$REGION/subscriptions/psltobt-sub \
+                                        --pubsublite.bigtable.streaming.checkpoint.location="gs://dataproc-templates/integration-testing/psltobt/checkpoint" \
+                                        --pubsublite.bigtable.output.project=$GCP_PROJECT \
+                                        --pubsublite.bigtable.output.instance="psltobt" \
+                                        --pubsublite.bigtable.output.table="output_table"
+                                    '''
```
[the remark this answers is not in the document]
Comment (tanyarw) [reply python/.ci/Jenkinsfile:481]: That's surprising. I've set default value to 60 [here](https://github.com/GoogleCloudPlatform/dataproc-templates/blob/7edb7db324bf950d665ccfde1b9820c5fb33a384/python/dataproc_templates/pubsublite/pubsublite_to_bigtable.py#:~:text=dest%3Dconstants.PUBSUBLITE_BIGTABLE_STREAMING_TIMEOUT,default%3D60%2C)

File: python/dataproc_templates/pubsublite/README.md [hunk 3]
```diff
@@ -55,3 +55,144 @@ export REGION=us-central1
...
+        {
+            "columnfamily":"date",
+            "columnname":"month",
+            "columnvalue":"March"
+        }
+    ] 
+}
+'''
+```
+
```
[the remark this answers is not in the document]
Comment (tanyarw) [reply python/dataproc_templates/pubsublite/README.md:177]: nice one!

File: python/dataproc_templates/pubsublite/pubsublite_to_bigtable.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+            self.write_to_bigtable(batch_df, args, logger)
```
[the remark this answers is not in the document]
Comment (tanyarw) [reply python/dataproc_templates/pubsublite/pubsublite_to_bigtable.py]: [Updated](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/725/commits/3722b7281e80abd24440418b52e29b1a13ce35a1)

File: python/test/pubsublite/test_pubsublite_to_bigtable.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        assert parsed_args["pubsublite.bigtable.streaming.checkpoint.path"] == "gs://temp-bucket/checkpoint"
+        assert parsed_args["pubsublite.bigtable.output.project"] == "my-project"
+        assert parsed_args["pubsublite.bigtable.output.instance"] == "bt-instance1"
+        assert parsed_args["pubsublite.bigtable.output.table"] == "output_table"
+        assert parsed_args["pubsublite.bigtable.output.column.families"] == "cf1, cf2, cf3"
+        assert parsed_args["pubsublite.bigtable.output.max.versions"] == 3
+
+    @mock.patch.object(pyspark.sql, 'SparkSession')
+    @mock.patch.object(pyspark.sql, 'DataFrame')
+    def test_run_pass_args2(self, mock_spark_session, mock_df):
```
[the remark this answers is not in the document]
Comment (tanyarw) [reply python/test/pubsublite/test_pubsublite_to_bigtable.py]: [Updated](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/725/commits/d9fb442c029bd041ec6e13d95a9d896139f18ebc)

File: python/.ci/Jenkinsfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (NiloFreitas) [reply python/.ci/Jenkinsfile:481]: Indeed. But when I tested without directly passing the timeout argument, I could not get the data to be read from pub/sub and written to Bigtable

File: python/dataproc_templates/pubsublite/pubsublite_to_bigtable.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NiloFreitas) [reply python/dataproc_templates/pubsublite/pubsublite_to_bigtable.py:219]: Oh got it

File: python/.ci/Jenkinsfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (tanyarw) [reply python/.ci/Jenkinsfile:481]: I see, will add it even so.

File: python/dataproc_templates/pubsublite/README.md [hunk 6]
```diff
@@ -55,3 +55,144 @@ export REGION=us-central1
...
+export SUBNET="projects/my-project/regions/us-west1/subnetworks/test-subnet"
+
+./bin/start.sh \
+-- --template=PUBSUBLITETOBIGTABLE \
+--pubsublite.bigtable.subscription.path="projects/$GCP_PROJECT/locations/$REGION/subscriptions/psltobt-sub" \
+--pubsublite.bigtable.streaming.checkpoint.location="gs://temp-bucket/checkpoint" \
+--pubsublite.bigtable.output.project="my-project" \
+--pubsublite.bigtable.output.instance="bt-instance-1" \
+--pubsublite.bigtable.output.table="output_table"
+```
```
[the remark this answers is not in the document]
Comment (tanyarw) [reply python/dataproc_templates/pubsublite/README.md:199]: [Done](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/725/commits/e433e361d9ae75cf135497641b5abbbd49410428)

File: python/.ci/Jenkinsfile [hunk 7]
```diff
@@ -430,6 +430,58 @@ pipeline {
...
+                                                    "columnvalue":"Karnataka"
+                                                },
+                                                {
+                                                    "columnfamily":"date",
+                                                    "columnname":"year",
+                                                    "columnvalue":"2023"
+                                                }
+                                            ] 
+                                        }
+                                        \'\'\'
```
Comment (vanshaj-bhatia) [inline python/.ci/Jenkinsfile:470]: Move this to the Reset/ Clean Resources stage on the top.

File: python/.ci/Jenkinsfile [hunk 8]
```diff
@@ -430,6 +430,58 @@ pipeline {
...
+    
+                                        cd python
+
+                                        ./bin/start.sh \
+                                        -- --template=PUBSUBLITETOBIGTABLE \
+                                        --pubsublite.bigtable.subscription.path=projects/$GCP_PROJECT/locations/$REGION/subscriptions/psltobt-sub \
+                                        --pubsublite.bigtable.streaming.checkpoint.location="gs://dataproc-templates/integration-testing/psltobt/checkpoint" \
+                                        --pubsublite.bigtable.output.project=$GCP_PROJECT \
+                                        --pubsublite.bigtable.output.instance="tanyawarrier-bt-1" \
```
Comment (vanshaj-bhatia) [inline python/.ci/Jenkinsfile:479]: Please avoid using your ldaps for integration testing. We are leveraging an existing bigtable instance for other java templates. You can use the same environment variable: $ENV_TEST_BIGTABLE_INSTANCE. Please rename the output table name as well.

**Review by vanshaj-bhatia — CHANGES_REQUESTED** [review]

File: python/.ci/Jenkinsfile (hunk 7 above)
Comment (tanyarw) [reply python/.ci/Jenkinsfile]: [Updated](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/725/commits/0f49e39d066668acc8b09591fafbfa7470416313)

**Review by NiloFreitas — APPROVED** [review]
LGTM

**Review by vanshaj-bhatia — APPROVED** [review]
LGTM

**Pull request merged**